### ISSUE TITLE
restic dies unexpectedly after some 8.5 GiB

### ISSUE THREAD

**Issue by hzensne1** [restic/restic#298]
Hi, Sorry it's me again with yet another problem. I have tried to backup data from one disk to another. Both disk are installed in my standard PC, running wheezy on an amd64 architecture. This is what I got:

scanned 21151 directories, 203251 files in 0:22
Getötet 3.71%  3.640 MiB/s  8.501 GiB / 229.098 GiB  10497 / 224402 items  0 errors  ETA 17:14:22

I started the same process again and got nearly the same output but obviously not with the same file (10497 vs 10495):

scanned 21151 directories, 203251 files in 0:22
Getötet 3.69%  3.736 MiB/s  8.454 GiB / 229.098 GiB  10495 / 224402 items  0 errors  ETA 16:47:58 

Observations:
restic process was eating a lot of CPU usage when it ran.
Destination (and source if of interest) disk have plenty of free space.

Should I try to run restic in debug mode and go through the output manually? I've not done that yet and wonder wether the output in the debugging file would be self-explanatory to me.
Thanks.

**Comment by fd0** [comment]
Thanks for reporting the issue. I suspect that restic was killed by the kernel OOM killer, could you look at `dmesg` or the kernel logs?

**Comment by hzensne1** [comment]
Wow that was a fast reply. Thanks. 
As usual I will be able to get dmesg output when I am in front of the machine again this evening.

**Comment by hzensne1** [comment]
You were right. The message said:
restic invoked oom-killer. So seems like an OS issue to tune some parameters first.

**Comment by fw42** [comment]
Not necessarily an OS issue, restic shouldn't use so much memory that it gets killed by the OS, imho. I think we should consider this a bug in restic. @fd0?

**Comment by fd0** [comment]
That depends on the amount of RAM that was needed ;)

**Comment by hzensne1** [comment]
Ok, I have added swapspace to my system and I am waiting for restic to reach the critical point again.
In the meantime let me suggest: Rather than having a clean memory management I - as a user - would like to have other functionality prioritized. E.g. deleting snapshots between two backups.

As far as the oom thingy is concerned, a simple pre-backup command line message would suffice, suggesting to add more swap space (or RAM) when restic detects less than 512MiB at startup. This may sound a bit weird but if you never heard of oom before (like me, using Linux for more than a decade now) this would be a good starting point for investigations on possible errors.

I know you all have limited resources. But you already have done a great job introducing restic. And reasonable features are always convincing to users who doubt if they should change their backup program/strategy.

By the way, restic has now processed item 10622 so it has overcome the critical point and progresses.
Thanks

**Comment by fd0** [comment]
Regarding the memory usage: We should add a check how much memory is available and balance this with the concurrency that restic employs (because that's the main reason restic needs more ram: more data blobs are processed in parallel). A warning message is also a good idea. I don't have time today, maybe somebody could add a new issue about memory/concurrency management (with an optional warning).

**Comment by fd0** [comment]
I'll consider this a bug.

**Comment by cowai** [comment]
I want to use restic for backuping tens of terabytes with millions of files. This is improbable to have working with the current restic behaviour, right? I would need lots of RAM wouldn't I?

**Comment by fd0** [comment]
restic hasn't been tested with so much data, please give it a try and report all issues you find, so we can fix them.

**Comment by hzensne1** [comment]
@cowai 
See my post above where I tried to backup 229 GiB in 21151 directories. The given ETA was 16:47:58 but I have killed the process just five minutes ago because it still was not finished (after more than 2 days). I am now backing up the subdirectories. Folder by folder really.

**Comment by fd0** [comment]
Wow, interesting. I need to test this myself.

**Comment by hzensne1** [comment]
@fd0 
I made a strange observation that I would like to share. I have started three parallel processes on three different folders. Two of the restic outputs showed a reasonable ETA of some 15 minutes. But the third folder went up to 99:50 hours. This folder basically contains jpegs. The other two folders mostly "normal" files.
Could it be that your compression+encryption procedure has problems compressing files that already have compression (like jpegs)?

**Comment by fd0** [comment]
The observation is interesting, I don't have any idea what the cause on that is. At the moment, restic doesn't do any compression, only deduplication and encryption. For a folder of JPEG files, I expect that basically restic does not find any duplicate chunks at all, so all chunks are new and need to be stored.

**Comment by maxigas** [comment]
I just started a backup to a freshly initialised repository of 1.224 TiB. It goes over the local network (NFS). Based on the issues of restic I assume it will not finish, but I will report on the progress/failure:

[19:57] 0.46%  4.946 MiB/s  5.781 GiB / 1.224 TiB  1650 / 17740... ETA 71:45:44

From a system administration point of view I consider it essential that backup software handle large amounts of data. By the way, attic ran for around a week and then crashed when it ran out of memory (with mmap failure), and there was no issue about it that I found. I'll try to report the issue too but need to test more...

**Comment by fd0** [comment]
I'd be very interested in your report and any crashed/bugs you might experience. Restic is a new project developed from scratch, I expect to fix many bugs until the 1.0.

**Comment by maxigas** [comment]
Restarted the backup from scratch after a power cut. Here is some performance data and comments after 10 hours into the backup:

`restic` progress report:

```
scanned 13051 directories, 164356 files in 0:18
[9:32:32] 14.28%  5.334 MiB/s  178.952 GiB / 1.224 TiB  40184 /... ETA 57:18:06
```

Good. Throughput is around 5MiB/s since the beginning, where the bottleneck is not restic but the network or the target NAS. Progress indicators appear to be consistent through time, and look more or less realistic to me.

`top` (after hittin "e" and "E" to display memory figures in megabytes):

```
top - 01:33:20 up 11:05,  2 users,  load average: 13.30, 12.33, 12.13
PID USER      PR  NI    VIRT    RES    SHR S  %CPU %MEM     TIME+ COMMAND    
2206 backup      20   0 1368.7m 462.1m   1.0m S 348.6 45.9   1289:45 restic
```

Problematic. Machine has 1G RAM, 2G swap, 4 CPUs (Intel Xeon 2.66GHz). As far as I see the memory usage of restic is consistently rising in tandem with the amount of data transferred. In fact it seems to me that even the CPU usage of restic is consistently rising in tandem with the amount of data transferred, which is harder to imagine. I don't understand well enough to reason about the root causes of such behaviour, but it does look like a bug.

Above tendency is consistent with hzensne1's initial observation starting this thread, e.g.: "restic process was eating a lot of CPU usage when it ran."

`free -m` output:

```
total       used       free     shared    buffers     cached
Mem:          1007        919         87          0          2        390
-/+ buffers/cache:        526        481
Swap:         2047        212       1835
```

**Comment by maxigas** [comment]
Next report: after another ten or so hours the memory as well as the CPU usage consistently increases:

Restic progress report:

```
scanned 13051 directories, 164356 files in 0:18
[21:23:21] 34.54%  5.758 MiB/s  433.005 GiB / 1.224 TiB  131020... ETA 40:31:58
```

Top:

```
top - 13:27:33 up 22:59,  2 users,  load average: 17.82, 14.33, 13.01
PID USER      PR  NI    VIRT    RES    SHR S  %CPU %MEM     TIME+ COMMAND    
2206 backup      20   0 1642.1m 686.6m   0.0m S  36.9 68.2   2980:20 restic
```

It is plain to see that load is going up -- this looks like it is heading to the hands of OOM killer.

**Comment by fd0** [comment]
I'm aware that the bookkeeping needed for the list of known data chunks (from the deduplication) is linear in memory usage. Thanks for your report, I think we need to investigate why restic seems to use more CPU time.

**Comment by maxigas** [comment]
Well, our infrastructure is obviously not sound enough to make very scientific tests like this... after another power cut and restic restart from scratch, now the backup is almost finished (~90%). The observations about rising memory and CPU usage still stand, as far as I am concerned, but it is hard to put a finger on the hard numbers to demonstrate, so I am not as sure as before.

One thing that could interfere with the linear rise of CPU is that there were several interruptions in the network connectivitiy (which both nfs and restic survived!).  Maybe -- just maybe -- restic CPU usage is rising while data is transferred but somehow recovered while blocking on a network transfer?  Well, that is just an idea.

The important point is that while still using considerable resources, it looks like this time the backup will actually finish (fingers crossed):

Restic progress report:

```
[65:09:11] 89.67%  4.907 MiB/s  1.098 TiB / 1.224 TiB  156087 / ... ETA 7:30:19
```

Top:

```
top - 14:53:02 up 3 days, 20:06,  2 users,  load average: 14.97, 14.92, 14.42
PID USER      PR  NI    VIRT    RES    SHR S  %CPU %MEM     TIME+ COMMAND     
5109 backup      20   0 1674.3m 792.4m   0.0m R 227.1 78.7   8124:31 restic  
```

hth & brb

**Comment by fd0** [comment]
Thanks for the report. Btw, I've finished work on #310 and will probably merge this branch in the evening. From there on restic regularly takes notes on what blobs have already been stored in the backend, so that backups can be resumed.

**Comment by maxigas** [comment]
Cool! Resuming will make it easier to backup a lot of data in volatile infrastructures like ours. :)

In the meantime, the backup finished without crashing the computer and I am now verifying it.

Thanks for writing restic.

**Comment by fd0** [comment]
Thanks for reporting your issue ;) this really helps in getting bugs ironed out.

**Comment by fd0** [comment]
Did you have a chance to try this again? Otherwise I'll close the issue for now, as people have succesfully created much larger repositories.

**Comment by fd0** [comment]
As announced this issue is closed now. Feel free to add a comment if you disagree.
